libraries/standard/http/src/http_client.c

```diff
@@ -247,40 +307,203 @@ static HTTPStatus_t _sendHttpBody( const HTTPTransportInterface_t * pTransport,
+    {
+        if( totalReceived == responseBufferLen )
         {
-            IotLogErrorWithArgs( "Error in sending the HTTP body over the "
-                                 "transport interface. Transport status %d.",
-                                 transportStatus );
-            returnStatus = HTTP_NETWORK_ERROR;
+            IotLogErrorWithArgs( "Response is too large for the response buffer"
+                                 ": responseBufferLen=%d",
+                                 responseBufferLen );
+            returnStatus = HTTP_INSUFFICIENT_MEMORY;
         }
-        else if( transportStatus != reqBodyBufLen )
+        else
         {
-            IotLogErrorWithArgs( "Failure in sending HTTP headers: Transport layer "
-                                 "did not send the required bytes: Required Bytes=%d"
-                                 ", Sent Bytes=%d.",
-                                 transportStatus );
-            returnStatus = HTTP_NETWORK_ERROR;
+            IotLogErrorWithArgs( "Partial response received: Transport recv "
+                                 "returned zero before the complete response: "
+                                 "Partial size=%d, Response buffer space "
+                                 "left=%d",
+                                 totalReceived,
+                                 responseBufferLen - totalReceived );
+            returnStatus = HTTP_PARTIAL_RESPONSE;
         }
     }
+    else
+    {
+        /* Empty else for MISRA 15.7 compliance. */
+    }
 
     return returnStatus;
 }
 
-/*-----------------------------------------------------------*/
-
-static HTTPStatus_t _receiveHttpResponse( const HTTPTransportInterface_t * pTransport,
-                                          HTTPResponse_t * pResponse )
+static HTTPStatus_t _receiveAndParseHttpResponse( const HTTPTransportInterface_t * pTransport,
+                                                  HTTPResponse_t * pResponse )
 {
-    /* TODO: Receive the HTTP response with parsing. */
-    return HTTP_SUCCESS;
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    size_t totalReceived = 0;
+    size_t currentReceived = 0;
+    HTTPParsingContext_t parsingContext = { 0 };
+
+    if( pResponse->pBuffer == NULL )
+    {
+        IotLogError( "Parameter check failed: pResponse->pBuffer is NULL." );
+        returnStatus = HTTP_INVALID_PARAMETER;
+    }
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        /* Initialize the parsing context. */
+        returnStatus = _HTTPClient_InitializeParsingContext( &parsingContext,
+                                                             pResponse->pHeaderParsingCallback );
+    }
+
+    /* While there are no errors in the transport recv or parsing, the response
+     * message is not finished, and there is room in the response buffer. */
+    while( ( returnStatus == HTTP_SUCCESS ) &&
+           ( parsingContext.state != HTTP_PARSING_COMPLETE ) &&
```

Comment:
Is `parsingContext.state != HTTP_PARSING_COMPLETE` the main indicator of unfinished response?